core/services/ocr2/plugins/ocr2keeper/evm/registry.go

```diff
@@ -0,0 +1,771 @@
+
+		if logs, err = r.poller.LogsWithSigs(
+			start,
+			end,
+			upkeepStateEvents,
+			r.addr,
+			pg.WithParentCtx(r.ctx),
+		); err != nil {
+			return fmt.Errorf("%w: %s", ErrLogReadFailure, err)
+		}
+
+		for _, log := range logs {
+			r.chLog <- log
+		}
+	}
+
+	r.mu.Lock()
+	r.lastPollBlock = end
+	r.mu.Unlock()
+
+	return nil
+}
+
+func (r *EvmRegistry) registerEvents(addr common.Address) error {
+	// Add log filters for the log poller so that it can poll and find the logs that
+	// we need
+	filterID, err := r.poller.RegisterFilter(logpoller.Filter{
+		EventSigs: append(upkeepStateEvents, upkeepActiveEvents...),
+		Addresses: []common.Address{addr},
+	})
+	if err != nil {
+		r.mu.Lock()
+		r.filterID = filterID
+		r.mu.Unlock()
+	}
+	return err
+}
+
+func (r *EvmRegistry) processUpkeepStateLog(l logpoller.Log) error {
+
+	rawLog := l.ToGethLog()
+	abilog, err := r.registry.ParseLog(rawLog)
+	if err != nil {
+		return err
+	}
+
+	switch l := abilog.(type) {
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepMigrated: // removes upkeep id and detail from registry
+		r.removeFromActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepCanceled:
+		r.removeFromActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepPaused:
+		r.removeFromActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepRegistered: // adds new upkeep id to registry
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepReceived: // adds multiple new upkeep ids to registry
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepUnpaused:
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryFundsAdded:
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepCheckDataUpdated:
+		r.addToActive(l.Id)
+	case *keeper_registry_wrapper2_0.KeeperRegistryUpkeepGasLimitSet:
+		r.addToActive(l.Id)
+	}
+
+	return nil
+}
+
+func (r *EvmRegistry) removeFromActive(id *big.Int) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	delete(r.active, id.Int64())
+}
+
+func (r *EvmRegistry) addToActive(id *big.Int) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, ok := r.active[id.Int64()]; !ok {
+		actives, err := r.getUpkeepConfigs(r.ctx, []*big.Int{id})
+		if err != nil {
+			return
+		}
+
+		if len(actives) != 1 {
+			return
+		}
+
+		r.active[id.Int64()] = actives[0]
+	}
+}
+
+func (r *EvmRegistry) buildCallOpts(ctx context.Context, block *big.Int) (*bind.CallOpts, error) {
+	opts := bind.CallOpts{
+		Context:     ctx,
+		BlockNumber: nil,
+	}
+
+	// integration tests don't support pending state because the simulated
+	// backend doesn't support it yet. the following commented code shoule be
+	// used when pending state support is added.
+	/*
+		if block == nil || block.Int64() == 0 {
+			opts.Pending = true
+		} else if block.Int64() == r.LatestBlock() {
+			opts.Pending = true
+		} else {
+			opts.BlockNumber = new(big.Int).Add(block, big.NewInt(1))
+		}
+	*/
```

Comment:
Can you link a ticket here. There are more nuances than just integration test support for pending block that we should document